## Fraud Risk Profiling & Anomaly Detection in Bank Accounts ##
##### Business Question:
Which account and transaction characteristics signal fraudulent account openings, and how can a fraud team prioritize investigations?

A fraud detection analysis that helps a bank's fraud team understand:
- **Who** is committing fraud (what do fraudulent applicants look like?)
- **How** they behave differently from legitimate customers
- **Where** the bank should focus its investigation efforts

Each record in the dataset represents a single transaction or application event, enriched with customer, behavioral, and identity-related attributes.

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/Base.csv")
df.head()

,fraud_bool,income,name_email_similarity,prev_address_months_count,current_address_months_count,customer_age,days_since_request,intended_balcon_amount,payment_type,zip_count_4w,...,has_other_cards,proposed_credit_limit,foreign_request,source,session_length_in_minutes,device_os,keep_alive_session,device_distinct_emails_8w,device_fraud_count,month
0,0,0.3,0.986506,-1,25,40,0.006735,102.453711,AA,1059,...,0,1500.0,0,INTERNET,16.224843,linux,1,1,0,0
1,0,0.8,0.617426,-1,89,20,0.010095,-0.849551,AD,1658,...,0,1500.0,0,INTERNET,3.363854,other,1,1,0,0
2,0,0.8,0.996707,9,14,40,0.012316,-1.490386,AB,1095,...,0,200.0,0,INTERNET,22.730559,windows,0,1,0,0
3,0,0.6,0.475100,11,14,30,0.006991,-1.863101,AB,3483,...,0,200.0,0,INTERNET,15.215816,linux,1,1,0,0
4,0,0.9,0.842307,-1,29,40,5.742626,47.152498,AA,2339,...,0,200.0,0,INTERNET,3.743048,other,0,1,0,0


**Each row = one application / transaction event**

##### Transaction:
- fraud_bool - Fraud label (1 if fraud, 0 if legit)
- credit_risk_score - Internal score of application risk. Ranges between [−170, 389].
- proposed_credit_limit - Applicant’s proposed credit limit. Ranges between [190, 2100].
- intended_balcon_amount - Initial transferred amount for application. Ranges between [−16, 113].
- payment_type - Credit payment plan type. 5 possible (annonymized) values.
- foreign_request - If origin country of request is different from bank’s country (0 for same country, 1 for different country).
- bank_branch_count_8w - Number of total applications in the selected bank branch in last 8 weeks. Ranges between [0, 2385].

##### Customer:
- income - annual income of the applicant in quantiles. Ranges between [0, 1]
- customer_age - Applicant’s age in bins per decade (e.g, 20-29 is represented as 20).
- prev_address_months_count - Number of months in previous registered address of the applicant, i.e. the applicant’s previous residence, if applicable. Ranges between [−1, 383] months (-1 is a missing value).
- current_address_months_count - Months in currently registered address of the applicant. Ranges between [−1, 428] months (-1 is a missing value).
- housing_status - Current residential status for applicant. 7 possible (annonymized) values.
- employment_status - Employment status of the applicant. 7 possible (annonymized) values.
- has_other_cards - If applicant has other cards from the same banking company (0 for no other cards, 1 for other cards). 

##### Session:
- device_os - Operative system of device that made request. Possible values are: Windows, Macintosh, Linux, X11, or other.
- source - Online source of application. Either browser(INTERNET) or mobile app (APP).
- session_length_in_minutes - Length of user session in banking website in minutes. Ranges between [−1, 86] minutes
- keep_alive_session - User option on session logout (0 for not kept alive, 1 for kept alive).
- device_distinct_emails_8w - Number of distinct emails in banking website from the used device in last 8 weeks. Ranges between [-1, 2] (-1 is a missing value).
- velocity_6h - Velocity of total applications made in last 6 hours i.e., average number of applications per hour in the last 6 hours. Ranges between [−211, 24763].
- velocity_24h - Velocity of total applications made in last 24 hours i.e., average number of applications per hour in the last 24 hours. Ranges between [1329, 9527].
- velocity_4w - Velocity of total applications made in last 4 weeks, i.e., average number of applications per hour in the last 4 weeks. Ranges between [2779, 7043].

##### Identity
- name_email_similarity - Metric of similarity between email and applicant’s name. Higher values represent higher similarity. Ranges between [0, 1].
- email_is_free - Domain of application email (0 for paid, 1 for free).
- phone_home_valid - Validity of provided home phone (0 for invalid, 1 for valid).
- phone_mobile_valid - Validity of provided mobile phone (0 for invalid, 1 for valid).
- date_of_birth_distinct_emails_4w - Number of different emails used with the same date of birth (last 4 weeks). Ranges between [0, 39].


##### Other feautures:
- days_since_request - Number of days passed since application was done. Ranges between [0, 78] days.
- bank_months_count - How old is previous account (if held) in months. Ranges between [−1, 31] months (-1 is a missing value).
- zip_count_4w - Number of applications within same zip code in last 4 weeks. Ranges between [1, 5767].
- month - Month where the application was made. Ranges between [0, 7].
- device_fraud_count - Number of fraudulent applications with used device. Ranges between [0, 1].

In [2]:
df.shape

(1000000, 32)

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 32 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   fraud_bool                        1000000 non-null  int64  
 1   income                            1000000 non-null  float64
 2   name_email_similarity             1000000 non-null  float64
 3   prev_address_months_count         1000000 non-null  int64  
 4   current_address_months_count      1000000 non-null  int64  
 5   customer_age                      1000000 non-null  int64  
 6   days_since_request                1000000 non-null  float64
 7   intended_balcon_amount            1000000 non-null  float64
 8   payment_type                      1000000 non-null  str    
 9   zip_count_4w                      1000000 non-null  int64  
 10  velocity_6h                       1000000 non-null  float64
 11  velocity_24h                      1000000 non-nul

In [4]:
df.nunique()

fraud_bool                               2
income                                   9
name_email_similarity               998861
prev_address_months_count              374
current_address_months_count           423
customer_age                             9
days_since_request                  989330
intended_balcon_amount              994971
payment_type                             5
zip_count_4w                          6306
velocity_6h                         998687
velocity_24h                        998940
velocity_4w                         998318
bank_branch_count_8w                  2326
date_of_birth_distinct_emails_4w        40
employment_status                        7
credit_risk_score                      551
email_is_free                            2
housing_status                           7
phone_home_valid                         2
phone_mobile_valid                       2
bank_months_count                       33
has_other_cards                          2
proposed_cr

In [5]:
df['fraud_bool'].value_counts()

fraud_bool
0    988971
1     11029
Name: count, dtype: int64

In [22]:
df.describe()

,fraud_bool,income,name_email_similarity,prev_address_months_count,current_address_months_count,customer_age,intended_balcon_amount,bank_branch_count_8w,date_of_birth_distinct_emails_4w,credit_risk_score,email_is_free,phone_home_valid,phone_mobile_valid,has_other_cards,proposed_credit_limit,foreign_request,session_length_in_minutes,keep_alive_session
count,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000,1000000.000000
mean,0.011029,0.562696,0.493694,16.718568,86.587867,33.689080,8.661499,184.361849,9.503544,130.989595,0.529886,0.417077,0.889676,0.222988,515.851010,0.025242,7.544940,0.576947
std,0.104438,0.290343,0.289125,44.046230,88.406599,12.025799,20.236155,459.625329,5.033792,69.681812,0.499106,0.493076,0.313293,0.416251,487.559902,0.156859,8.033106,0.494044
min,0.000000,0.100000,0.000001,-1.000000,-1.000000,10.000000,-15.530555,0.000000,0.000000,-170.000000,0.000000,0.000000,0.000000,0.000000,190.000000,0.000000,-1.000000,0.000000
25%,0.000000,0.300000,0.225216,-1.000000,19.000000,20.000000,-1.181488,1.000000,6.000000,83.000000,0.000000,0.000000,1.000000,0.000000,200.000000,0.000000,3.103053,0.000000
50%,0.000000,0.600000,0.492153,-1.000000,52.000000,30.000000,-0.830507,9.000000,9.000000,122.000000,1.000000,0.000000,1.000000,0.000000,200.000000,0.000000,5.114321,1.000000
75%,0.000000,0.800000,0.755567,12.000000,130.000000,40.000000,4.984176,25.000000,13.000000,178.000000,1.000000,1.000000,1.000000,0.000000,500.000000,0.000000,8.866131,1.000000
max,1.000000,0.900000,0.999999,383.000000,428.000000,90.000000,112.956928,2385.000000,39.000000,389.000000,1.000000,1.000000,1.000000,1.000000,2100.000000,1.000000,85.899143,1.000000


## Feature Selection

Let's perform feature selection by removing low-variance and non-informative variables, keeping only those relevant for fraud risk analysis.
##### Identify Strong Signals of Fraud 
##### Numerical Features

In [6]:
pd.set_option('display.max_columns', None)
df.groupby("fraud_bool").mean(numeric_only=True)

,income,name_email_similarity,prev_address_months_count,current_address_months_count,customer_age,days_since_request,intended_balcon_amount,zip_count_4w,velocity_6h,velocity_24h,velocity_4w,bank_branch_count_8w,date_of_birth_distinct_emails_4w,credit_risk_score,email_is_free,phone_home_valid,phone_mobile_valid,bank_months_count,has_other_cards,proposed_credit_limit,foreign_request,session_length_in_minutes,keep_alive_session,device_distinct_emails_8w,device_fraud_count,month
fraud_bool,,,,,,,,,,,,,,,,,,,,,,,,,,
0,0.561313,0.494815,16.839647,86.273232,33.609125,1.025383,8.713907,1572.138693,5670.664988,4771.528849,4857.444566,184.923747,9.526521,130.469904,0.528423,0.418906,0.890112,10.843426,0.224533,512.303162,0.024962,7.537306,0.579571,1.017630,0.0,3.285582
1,0.686635,0.393161,5.861365,114.801161,40.858645,1.054615,3.962009,1622.311542,5183.913444,4613.138798,4755.844185,133.976426,7.443195,177.590353,0.661075,0.253060,0.850576,10.469580,0.084414,833.986762,0.050322,8.229520,0.341645,1.079427,0.0,3.565962


Insights:
- Fraudulent users tend to have higher income levels
- Fraud is more common among older customers
- Fraudsters have short previous address history but long current address (suspicious pattern)
- Fraud strongly associated with higher credit risk
- Fraudulent profiles request higher credit limits
- More mismatch = more fraud
- Fraudsters use free emails more often
- Fraudsters have less valid contact info
- Fraudulent users tend to have lower intended amounts
- Session length is slightly higher and keep-alive session is lower in fraud, which can mean possibly shorter or inconsistent sessions
- Fraudulent accounts may behave less consistently or differently over time
- Fraudsters are less likely to have multiple cards

Columns that doesn’t differentiate fraud:
- days_since_request
- zip_count_4w
- velocity_6h
- velocity_24h
- velocity_4w
- bank_months_count
- device_fraud_count
- month

##### Categorical Features

##### - Payment Type

In [7]:
pd.crosstab(df["payment_type"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
payment_type,,
AA,0.994718,0.005282
AB,0.988749,0.011251
AC,0.983302,0.016698
AD,0.989178,0.010822
AE,0.996540,0.003460


- Certain payment types (especially AC - 1.67%) are significantly more associated with fraud.

##### - Employment Status

In [8]:
pd.crosstab(df["employment_status"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
employment_status,,
CA,0.987814,0.012186
CB,0.993109,0.006891
CC,0.975316,0.024684
CD,0.996230,0.003770
CE,0.997664,0.002336
CF,0.998070,0.001930
CG,0.984547,0.015453


- Certain employment categories (especially CC - 2.47%) show much higher fraud risk.

##### - Housing Status

In [9]:
pd.crosstab(df["housing_status"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
housing_status,,
BA,0.962534,0.037466
BB,0.993992,0.006008
BC,0.993852,0.006148
BD,0.991361,0.008639
BE,0.996559,0.003441
BF,0.995806,0.004194
BG,0.996032,0.003968


- Housing status BA (3.75%) is a major fraud indicator.

##### - Source

In [10]:
pd.crosstab(df["source"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
source,,
INTERNET,0.989006,0.010994
TELEAPP,0.984109,0.015891


- Transactions initiated via TELEAPP show a slightly higher fraud rate compared to INTERNET, indicating potential channel-based risk differences.

##### - Device OS

In [11]:
pd.crosstab(df["device_os"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
device_os,,
linux,0.994845,0.005155
macintosh,0.986029,0.013971
other,0.994240,0.005760
windows,0.975306,0.024694
x11,0.988794,0.011206


- Users on Windows devices show significantly higher fraud rates compared to other operating systems.

##### Drop weak / useless signals
Features were selected based on their discriminatory power between fraudulent and legitimate cases, using statistical comparison of group means and fraud rates.

In [12]:
df = df.drop(columns=[
    "device_fraud_count", 
    "month", 
    "zip_count_4w", 
    "bank_months_count", 
    "days_since_request",  
    "velocity_6h", 
    "velocity_24h", 
    "velocity_4w"
])

##### Detailed analysis on some feautures

##### - Income

In [13]:
pd.crosstab(df["income"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
income,,
0.1,0.994227,0.005773
0.2,0.993684,0.006316
0.3,0.993351,0.006649
0.4,0.992663,0.007337
0.5,0.992051,0.007949
0.6,0.991221,0.008779
0.7,0.991181,0.008819
0.8,0.989076,0.010924
0.9,0.978362,0.021638


- Fraud increases with higher income levels, especially at the top range (0.9 - 2.16%).

##### - Credit Risk Score

In [14]:
pd.crosstab(
    pd.qcut(df["credit_risk_score"], 5, duplicates="drop"),
    df["fraud_bool"],
    normalize="index"
)

fraud_bool,0,1
credit_risk_score,,
"(-170.001, 73.0]",0.993882,0.006118
"(73.0, 107.0]",0.993279,0.006721
"(107.0, 141.0]",0.993470,0.006530
"(141.0, 192.0]",0.988637,0.011363
"(192.0, 389.0]",0.975299,0.024701


- Higher credit risk scores are strongly associated with fraud (2.47%).

##### - Intended Balcon Amount

In [15]:
df["intended_balcon_amount"].quantile([0.25, 0.5, 0.75])

0.25   -1.181488
0.50   -0.830507
0.75    4.984176
Name: intended_balcon_amount, dtype: float64

In [16]:
pd.crosstab(
    pd.qcut(df["intended_balcon_amount"], 3, duplicates="drop"),
    df["fraud_bool"],
    normalize="index"
)

fraud_bool,0,1
intended_balcon_amount,,
"(-15.532, -1.063]",0.988066,0.011934
"(-1.063, -0.516]",0.986314,0.013686
"(-0.516, 112.957]",0.992533,0.007467


- The intended transaction amount is an encoded feature. While absolute values are not directly interpretable, relative differences reveal that lower values are associated with higher fraud risk.

##### - Device Distinct Emails

In [17]:
pd.crosstab(df["device_distinct_emails_8w"], df["fraud_bool"], normalize="index")

fraud_bool,0,1
device_distinct_emails_8w,,
-1,0.988858,0.011142
0,0.975925,0.024075
1,0.989836,0.010164
2,0.959094,0.040906


- Devices associated with multiple email identities show significantly higher fraud rates, indicating potential account cycling or synthetic identity behavior.

In [18]:
df["device_email_flag"] = df["device_distinct_emails_8w"].map({
    -1: "unknown",
    0: "no_email",
    1: "normal",
    2: "multiple"
})

In [19]:
df = df.drop(columns=["device_distinct_emails_8w"])

##### - Date of Birth of Distinct Emails

In [20]:
pd.crosstab(
    df["date_of_birth_distinct_emails_4w"],
    df["fraud_bool"],
    normalize="index"
)

fraud_bool,0,1
date_of_birth_distinct_emails_4w,,
0,0.952096,0.047904
1,0.957988,0.042012
2,0.967634,0.032366
3,0.981098,0.018902
4,0.987594,0.012406
5,0.988292,0.011708
6,0.989211,0.010789
7,0.988449,0.011551
8,0.987043,0.012957


- Customers with fewer email associations per date of birth show significantly higher fraud rates, suggesting the use of newly created or isolated identities in fraudulent activity.

##### CSV files for sql

In [20]:
df["transaction_id"] = range(1, len(df)+1)
df["customer_id"] = range(1, len(df)+1)
df["session_id"] = range(1, len(df)+1)
df["identity_id"] = range(1, len(df)+1)

In [21]:
transaction_df = df[[
    "transaction_id",
    "customer_id",
    "identity_id",
    "session_id",
    "fraud_bool",
    "credit_risk_score",
    "proposed_credit_limit",
    "intended_balcon_amount",
    "payment_type",
    "foreign_request",
    "bank_branch_count_8w"
]]

customer_df = df[[
    "customer_id",
    "income",
    "customer_age",
    "prev_address_months_count",
    "current_address_months_count",
    "housing_status",
    "employment_status",
    "has_other_cards"
]]

session_df = df[[
    "session_id",
    "device_os",
    "source",
    "session_length_in_minutes",
    "keep_alive_session",
    "device_email_flag"
]]

identity_df = df[[
    "identity_id",
    "name_email_similarity",
    "email_is_free",
    "phone_home_valid",
    "phone_mobile_valid",
    "date_of_birth_distinct_emails_4w"
]]

In [22]:
customer_df.to_csv("../data/clean/customer.csv", index=False)

In [24]:
from sqlalchemy import create_engine
import getpass

password = getpass.getpass("Please, input your SQL password: ")

Please, input your SQL password:  ········


In [25]:
database_name = "fraud_analysis"
connection_string = 'mysql+pymysql://root:' + password + '@localhost/'+ database_name
engine = create_engine(connection_string)
engine

Engine(mysql+pymysql://root:***@localhost/fraud_analysis)

In [26]:
engine.connect()

In [27]:
customer_df.to_sql(
    "customer",
    engine,
    if_exists="append",
    index=False,
    chunksize=10000
)

1000000

In [31]:
session_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 6 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   session_id                 1000000 non-null  int64  
 1   device_os                  1000000 non-null  str    
 2   source                     1000000 non-null  str    
 3   session_length_in_minutes  1000000 non-null  float64
 4   keep_alive_session         1000000 non-null  int64  
 5   device_email_flag          1000000 non-null  str    
dtypes: float64(1), int64(2), str(3)
memory usage: 45.8 MB


In [37]:
session_df.to_sql(
    "session",
    engine,
    if_exists="append",
    index=False,
    chunksize=10000
)

1000000

In [38]:
identity_df.to_sql(
    "identity",
    engine,
    if_exists="append",
    index=False,
    chunksize=10000
)

1000000

In [40]:
transaction_df.to_sql(
    "transaction",
    engine,
    if_exists="append",
    index=False,
    chunksize=10000
)

1000000